server/src/main/java/org/elasticsearch/rest/ChunkedZipResponse.java

```diff
@@ -0,0 +1,485 @@
+     * listener.
+     *
+     * @param entryName The name of the entry in the response zip file.
+     * @param listener  A listener which is completed when the entry has been completely processed: either fully sent, or else the request
+     *                  was cancelled and the response will not be used any further. If the returned entry listener is completed
+     *                  exceptionally then the exception is passed to {@code listener}, otherwise this listener is completed successfully.
+     */
+    public ActionListener<ChunkedRestResponseBodyPart> newEntryListener(String entryName, ActionListener<Void> listener) {
+        if (listenersRefs.tryIncRef()) {
+            final var zipEntry = new ZipEntry(filename + "/" + entryName);
+            return ActionListener.assertOnce(ActionListener.releaseAfter(listener.delegateFailureAndWrap((l, firstBodyPart) -> {
+                if (firstBodyPart == null) {
+                    l.onResponse(null);
+                } else {
+                    enqueueEntry(zipEntry, firstBodyPart, l);
+                }
+            }), listenersRefs::decRef));
+        } else {
+            assert false : "already closed";
+            throw new AlreadyClosedException("response already closed");
+        }
+    }
+
+    /**
+     * A zip file entry which is ready for transmission, to be stored in {@link #entryQueue}.
+     *
+     * @param zipEntry      The entry metadata, to be written in its header.
+     * @param firstBodyPart The first part of the entry body. Subsequent parts, if present, come from
+     *                      {@link ChunkedRestResponseBodyPart#getNextPart}.
+     * @param releasable    A resource to release when this entry has been fully transmitted, or is no longer required because the
+     *                      transmission was cancelled.
+     */
+    private record ChunkedZipEntry(ZipEntry zipEntry, ChunkedRestResponseBodyPart firstBodyPart, Releasable releasable) {}
+
+    /**
+     * Queue of entries that are ready for transmission.
+     */
+    private final Queue<ChunkedZipEntry> entryQueue = new LinkedBlockingQueue<>();
+
+    /**
+     * Upper bound on the number of entries in the queue, atomically modified to ensure there's only one thread processing queue entries at
+     * once.
+     */
+    private final AtomicInteger queueLength = new AtomicInteger();
+
+    /**
+     * Ref-counting for access to the queue, to avoid clearing the queue on abort concurrently with an entry being sent.
+     */
+    private final RefCounted queueRefs = AbstractRefCounted.of(this::drainQueue);
+
+    /**
+     * Flag to indicate if the request has been aborted, at which point we should stop enqueueing more entries and promptly clean up the
+     * ones being sent.
+     */
+    private final AtomicBoolean isRestResponseFinished = new AtomicBoolean();
+
+    private boolean tryAcquireQueueRef() {
+        return isRestResponseFinished.get() == false && queueRefs.tryIncRef();
```

Comment:
Yeah that's fine, the `isRestResponseFinished` check is only really there to prevent starvation of the drain-and-close path if there are enough other things concurrently acquiring refs.